### PR TITLE
STORM-2937: Overwrite latest storm-kafka-client 1.x-branch into 1.0.x-branch

### PR DESCRIPTION
**Pull request by erikdw**
Mimics @HeartSaVioR 's work in STORM-2936 and PR #2549, but applies back to the more divergent 1.0.x-branch.  Big thanks to @HeartSaVioR , @srdo, @hmcl for guidance and @jessicalhartog & @srishtyagrawal for pairing to figure out all the odds and ends here.  And for the storm community's flexibility and willingness to address this kind of issue!

* origin commit sha (1.x-branch): 74ca795
  * That is the same as in STORM-2936 and PR #2549
* changed 2 things in external/storm-kafka-client/pom.xml:
   * module version had to be synced with 1.0.x-branch current version
   * storm-core dependency's scope changed from undefined `${provided.scope}` to `provided`
* updated storm's base pom.xml with the same kafka client version as 1.x-branch: 0.10.1.0
* overwriting requires a few somewhat isolated changes to storm-core
  * added InterfaceStability.java
  * overwrote Time.java with that from 1.x-branch
     * needed both for tests in storm-kafka-client, and for Time.nanoTime() invocation
  * backported changes to SlotTest.java to leverage Time.java changes
     * needed since startSimulatingAutoAdvanceOnSleep() was ditched in the Time.java changes.
  * backport changes to Fields.java to allow equality comparisons in tests
      * otherwise we received confusing test failures claiming 2 seemingly identical items to be different
  * backport changes to trident interface for partition handling, which were needed for KafkaTridentSpoutEmitter
* also copied back the docs

NOTE: there were other minor-ish things I spotted in 1.x-branch that I intentionally didn't fix for now (typos, ^M characters, trailing whitespace, etc.).  I didn't want this to entail any more changes than necessary.

For quick reference, here are the files outside of storm-kafka-client that have been modified or added:

```
% git diff origin/1.0.x-branch | grep '^diff' | awk '{print $3}' | grep -v '^external/storm-kafka-client/'
docs/index.md
docs/storm-kafka-client.md
external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/trident/OpaqueTridentEventHubEmitter.java
external/storm-kafka/src/jvm/org/apache/storm/kafka/trident/TridentKafkaEmitter.java
pom.xml
storm-core/src/jvm/org/apache/storm/annotation/InterfaceStability.java
storm-core/src/jvm/org/apache/storm/trident/spout/IOpaquePartitionedTridentSpout.java
storm-core/src/jvm/org/apache/storm/trident/spout/OpaquePartitionedTridentSpoutExecutor.java
storm-core/src/jvm/org/apache/storm/trident/topology/state/TransactionalState.java
storm-core/src/jvm/org/apache/storm/tuple/Fields.java
storm-core/src/jvm/org/apache/storm/utils/Time.java
storm-core/test/jvm/org/apache/storm/daemon/supervisor/SlotTest.java
```

### CODE REVIEW

**Comment by ptgoetz** [comment]
+1 successfully built for me and all tests passed.

As far as squashing commits my opinion has always been if we do squash, it should be done when merging. There are a lot of cases where separate commits make review easier (e.g. separating pure whitespace changes). If commits are squashed during review, you have to re-read all the changes.

**Comment by srdo** [comment]
+1

**Comment by hmcl** [comment]
@erikdw I agree that in this case where we are cherry-picking several commits it's very hard to avoid multiple commits, and probably we shouldn't even try to squash them because that will completely erase historical context and where things came from. Nevertheless, I would like to suggest that the all commit messages for this PR follow the convention[1] 

```
STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
 - Message which ideally should contain the message and SHA of the commit that was cherry-picked. IntelliJ, and git cherry-pick typically do that by default
``` 

Currently git log for this PR (truncated to be easy to read here) looks like this:

```
ad7360507 add storm-kafka-client doc from 1.x-branch, and link to it from index.md
0ddc41154 copy Time.java from 1.x-branch to allow use of nanoTime() in storm-kafka-client
5d9ad1eaa backport changes to Fields to allow it to be checked for equality in storm-kafka-client 
f758123b5 backport trident interface changes
e42ae4de8 update storm-kafka-client pom.xml and base pom.xml as necessary for using storm-
533ef7ad2 copied external/storm-kafka-client from 1.x-branch at SHA 74ca795 (this doesn't
9c8930036 Merge branch 'STORM-2918-1.0.x-merge' into 1.0.x-branch
4df41444b STORM-2918 Update Netty version
5b507eaf0 Merge branch 'STORM-2853-1.0.x-merge' into 1.0.x-branch
4841475a7 STORM-2853 Initialize tick tuple after initializing spouts/bolts
```

I think that it would be much easier to trace back the bulk change if the git log was something like:

```
ad7360507 STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
0ddc41154 STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
5d9ad1eaa STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
f758123b5 STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
e42ae4de8 STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
533ef7ad2 STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
9c8930036 Merge branch 'STORM-2918-1.0.x-merge' into 1.0.x-branch
4df41444b STORM-2918 Update Netty version
5b507eaf0 Merge branch 'STORM-2853-1.0.x-merge' into 1.0.x-branch
4841475a7 STORM-2853 Initialize tick tuple after initializing spouts/bolts
```

And, for instance, for SHA ad7360507 the message would be.

```
STORM-2937: Overwrite storm-kafka-client 1.x-branch into 1.0.x-branch
 - add storm-kafka-client doc from 1.x-branch, and link to it from index.md
      (cherry picked from commit SHA) - this would be optional if its a big pain, but I think git does it for you
``` 

Similarly for the other SHAs.

[1] - I don't necessarily mean Storm convention

**Comment by hmcl** [comment]
@ptgoetz, @srdo and I have been using the following approach for squashing the patches in storm-kafka-client, which I believe works well: 

- It is reasonable to submit a PR for review that has a several commits, typically done during development.
- During the PR review, changes addressing code review comments go in a new commit such that it is easy to verify if the code review comments are property addressed.
- Once the PR has a +1 from the reviewers and the changes are accepted to merge the creator of the PR will squash the commits.
- I would vote for having only one commit with the commit message starting either with STORM-1234: or MINOR:. I would favor not having commit messages starting with arbitrary messages. However, this is a very debatable subject and I am OK If we don't want to enforce that small commits  have to be prefixed by a token such as MINOR. I would say that significant patches should be required to have the JIRA number at the top.
- I have discussed previously, and I believe that it is not beneficial to to the community to merge into master PRs that have several commits that follow a pattern along the lines:

```
STORM-1234: Some storm feature
fix compilation issue
address code review
fix checkstyle
fix unit tests
...
```

I think the entire community would benefit from a more structured way of merging code into the main code line. Perhaps we could try again proposing some PR and commit guidelines for voting. What do you think?

**Comment by HeartSaVioR** [comment]
> The squashing should be done by the creator of the PR and not by the committer that is going to merge the code.

Sorry but I have been exposing disagreement for this and I still disagree. You're referencing other big projects as a reference of your opinion but they require merger to squash commits (with letting handy merge script do it instead), never make contributors unhappy with dealing with squashing. I have been requested squashing, and many contributors who are not familiar with struggled and ended up opening new pull request.

**Comment by HeartSaVioR** [comment]
Other than that I agree most of @hmcl proposed, which is along with the best practice for what other big projects are following. If we require (not optional) squashing commits, I don't think merge commit is needed: it will become just redundant.

**Comment by erikdw** [comment]
@hmcl @HeartSaVioR : perhaps we can take this offline?  (I admit to being guilty of starting this discussion in these PRs.)

I am not grasping the actual problems that you encounter that make squashing so important.  I have read everything you wrote here and in the other thread where we discussed it.  But I still cannot see what *exactly* is the problem for you, nor why all these other big projects have gone to this information-losing approach.  All commits of a PR *are* bunched together.  They can all be referenced as a group in `git revert` or `git cherry-pick`.   Please note I do appreciate that you are willing to allow for divergence from a strict "squash all the things" standard when it is valuable.

NOTE: I abhor super noisy PRs that merge tons of bug-fixy commits, especially since that encourages bad commit messages as people usually quickly write "fixed foo" with no details when they are iteratively committing.  So it's a bit uncomfortable to find myself defending *not* squashing.

**Comment by hmcl** [comment]
@HeartSaVioR I understand your point of not putting the burden on the contributors. I didn't really think that it could be a big hurdle for contributors. However, taking that into consideration it makes sense that the person merging the code  should also squashing the commits. However, in that scenario the merger would also be responsible for the commit message, which in my opinion is responsibility of the contributor.

**Comment by erikdw** [comment]
To be clear, I'm largely aligned with what you're both saying.  I might not understand what precise problem you face that makes not squashing such a problem, but I too would appreciate the aesthetic clarity of a cleaner git log than we currently have, *so long as* it doesn't lose valuable information.  i.e., the stance of *always* squashing *will* cause information loss in the git log.

**Comment by hmcl** [comment]
@erikdw I am not sure I follow when you wrote:

> All commits of a PR are bunched together. They can all be referenced as a group in git revert or git cherry-pick. 

In the first git log example I gave in my [earlier comment](https://github.com/apache/storm/pull/2550#issuecomment-363955298), how can one find out from git log (i.e. commit history) that SHAs ad7360507, 0ddc41154, 5d9ad1eaa, f758123b5, e42ae4de8, 533ef7ad2 all belong to JIRA STORM-2937.

Perhaps you mean it is possible to do it from GitHub?

**Comment by HeartSaVioR** [comment]
@erikdw @hmcl @ptgoetz 
First of all, I need to make this clear: I'm not claiming that this PR should be changed. I'm OK to merge this in as it is, and I'll do it once 24hrs policy is met.

What we want to discuss is the standard rule of Storm, hence it should not be resolved in offline discussion, but discussion thread in dev list should be initiated and discussion should take place in there. Standard rule should not be brought from decision among part of us.

While discussing we need to consider both: 1. the unit of commits 2. title/body of commit. Without having standard rule, everyone think differently about twos, contributors and even PMCs. Letting contributors handle any of two would fail because of that, and if we have standard rule, it is normally the thing to be automated which everyone is not bothered to handle it manually.

Please take a look at below commits:

https://github.com/apache/spark/commit/30295bf5a6754d0ae43334f7bf00e7a29ed0f1af
https://github.com/apache/zeppelin/commit/c6b73beae71f79b48d834dc751182ec45d244d69

above two projects are having template of PR, and use it for composing body of commit, which will guarantee commit title/body to conform to their standard rule. We may bother with filling the template, that's a possible disadvantage.

and other commits:

https://github.com/apache/kafka/commit/7c97b239a5261cf030984f71427287196e657d95

above project don't require filling template but body of PR is taken to commit body with additional information.

https://github.com/apache/hadoop/commit/f491f717e9ee6b75ad5cfca48da9c6297e94a8f7
https://github.com/apache/hbase/commit/3bb8daa60565ec2f7955352e52c2f6379176d8c6
... and more ...

above projects don't require filling template. Maybe commit body is came from commit of contributor. In this case most of commits are not having body of commit message.

**Comment by erikdw** [comment]
@hmcl : now we're going down the rathole that I knew existed if we didn't take this offline. 😄 

Before I get to the question of using command-line `git` to see the related commits, I must note my agreement that all commits should be prefixed with the same ID (e.g., `STORM-2937:`) and perhaps a common brief description [1].  That would provide a convention for seeing that the commits are all part of the same JIRA.  That *might* suffice for satisfying your question.

As for how I personally would enumerate the commits in the `git log` for this PR as you requested, I cannot show you *yet* because this PR hasn't been merged.  So if you'll excuse the reference to a separate repo, I'll show you how I enumerate the commits for a PR different in a different repo:
* https://github.com/mesos/storm/pull/233/commits

We can see that the commits there are:
* f9fa0f6
* 03030db

When I look at my "git log" for purposes of seeing the flow of history, I use`git l` [2].  This output looks like so:

```
* 4119581 - (base/master) avoid hardcoding storm-core version in storm-shim-1x/pom.xml, just use the property defined in the project's base pom.xml (2018-01-23 18:56:03 -0800) <Erik Weathers>
*   256d5f0 - Merge pull request #231 from JessicaLHartog/logviewer-sidecar-flag (2017-12-20 11:40:06 -0800) <JessicaLHartog>
|\  
| * 18832dd - Conditionally perform logviewer sidecar creation depending on configuration. (2017-12-13 15:28:21 -0800) <Jessica Hartog>
| * 6dd4973 - Rename now defunct  configuration parameter (2017-12-13 14:46:29 -0800) <Jessica Hartog>
* |   d2cedb8 - Merge pull request #233 (2017-12-20 11:30:34 -0800) <JessicaLHartog>
|\ \  
| * | 03030db - Update protobuf version to be compatable with newer versions of Mesos (2017-12-18 15:16:37 -0800) <Jessica Hartog>
| * | f9fa0f6 - Updating storm version (2017-12-18 15:16:37 -0800) <Jessica Hartog>
|/ /  
* |   7a1e50c - Merge pull request #232 from JessicaLHartog/issue-222 (2017-12-18 15:12:29 -0800) <JessicaLHartog>
|\ \  
| |/  
|/|   
```

I acknowledge that it's a bit hard to parse visually at first!  But it shows me that there are 2 commits that went into the merge of PR #233:
* 03030db
* f9fa0f6

Then you can put those commits in a sequence when you run various git commands.

Again, I **do** agree that it's best practice to squash.  I'm just trying to argue that it shouldn't be a blanket unwavering policy with no exceptions.  Otherwise I believe we're choosing ease of commit stewardship over ease of reading and maintaining the actual code.

[1] I kind of dislike having the *exact same* headline with nothing unique for each commit because it forces you to look more deeply to see what each commit does (e.g., open each commit separately in GitHub, or use git show for each commit, or use git log and scroll to the commits you are looking at).  That is a minor nit with the convention you proposed though. 

[2] Definition of my `git l` alias

```
% cat ~/.gitconfig
...
[alias]
...
        l = log --date=iso-local --color --graph --pretty=format:'%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cd) %C(bold blue)<%an>%Creset' --abbrev-commit
```

**Comment by erikdw** [comment]
> First of all, I need to make this clear: I'm not claiming that this PR should be changed. I'm OK to merge this in as it is, and I'll do it once 24hrs policy is met.

Let me first add just a bit more info to my PR's commit messages per @hmcl's suggestions.

> What we want to discuss is the standard rule of Storm, hence it should not be resolved in offline discussion, but discussion thread in dev list should be initiated and discussion should take place in there. Standard rule should not be brought from decision among part of us.

Totally agree.  Sorry that my use of "offline" was unclear.  I meant offline from *this* channel of discussion.  In a dev-list email thread sounds apropos.  [1]  So maybe we start a thread there?

[1] Except that the format of the mails on the list are pretty ugly, and thus it's harder to cleanly express certain things.  e.g., my previous message there would not be nearly as clean in the email list.  Not sure why the lists seemingly don't support HTML emails.

**Comment by HeartSaVioR** [comment]
> Let me first add just a bit more info to my PR's commit messages per @hmcl's suggestions.

OK sure. Please let me know when you finished.

> Except that the format of the mails on the list are pretty ugly, and thus it's harder to cleanly express certain things.

Totally agree. Maybe you could use Storm wiki or gist and link as a complement.

**Comment by HeartSaVioR** [comment]
We may also discuss about rebase (along with squash in this case) vs merge, and I'm in favor of rebase, because we have huge chance for PR to be reviewed and merged not immediately. In this case the merge commit and actual commits would be far from the `git log` and it would be not easy to follow the commits even commit message of actual commits are well written.

Yes I'll mention whole the things again when we start discussion from dev list. Let's make 1.x version line releases be finished successfully first.

**Comment by erikdw** [comment]
@HeartSaVioR : I just updated all the commit messages in-line with @hmcl's proposal.  I also noticed that I missed the addition of the `serialVersionUID` to Fields.java so I put that in and rebased.  That's the only non-message change in the PR.

**Comment by hmcl** [comment]
+1.

Thanks @erikdw, it's really good. Thanks also for the discussion. Looking forward to continue it in the dev list, and hopefully agreeing on a good solution.
